### ISSUE TITLE
Bug: Getting 'no-unused-vars' for parameter of a function type in tsx

### ISSUE THREAD

**Issue by nicholasq** [eslint/eslint#16958]
### Environment

Node version: v18.12.1
npm version: v8.19.2
Local ESLint version: v8.35.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 22.3.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

I am using the default configuration from [Electron react boilerplate](https://github.com/electron-react-boilerplate/electron-react-boilerplate) which uses the airbnb eslint configuration.

</details>

```js
// MenuBar.tsx
export default function MenuBar({ items }: { items: MenuBarProps }) {
  const { syncing, toggleSync } = useSyncData((syncStatus: boolean) => {
    console.log('syncing', status);
  });
  return (<div></div>);
}


// sync-hook.ts
export type SyncCallBack = (syncStatus: boolean) => void;

export default function useSyncData(cb: SyncCallBack) {
  const [syncing, setSyncing] = useState(false);
  function handleSync() {
    if (!syncing) {
      setSyncing(true);
      cb(true);
    } else {
      setSyncing(false);
      cb(false);
    }
  }
   ... lots of code removed for brevity
}

```

### What did you expect to happen?

I expect to not have any lint errors or warnings.

### What actually happened?

Eslint throws this error:
`4:29  error  'syncStatus' is defined but never used  no-unused-vars`

But you can see that `syncStatus` is indeed being used.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @nicholasq, thanks for the issue!

For TypeScript code, you should use [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars) rule instead of the core `no-unused-vars` rule.
